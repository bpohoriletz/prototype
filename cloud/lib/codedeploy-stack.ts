import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';

import { createCdApplication } from './aws-cdk-kit/codedeploy/app';
import { createCodedeployRole } from './aws-cdk-kit/codedeploy/role';
import { createPublicDeploymentGroup } from './aws-cdk-kit/codedeploy/deployment-group';
import { createEc2InstanceProfile } from './aws-cdk-kit/roles/ec2-profile';
// import { createGithubCliRole } from './aws-cdk-kit/iam/role';
import { createTempPrivateBucket } from './aws-cdk-kit/s3/bucket';
import PetVpcBuilder from './aws-cdk-kit/ec2/vpc-builders/pet';
import VpcDirector from './aws-cdk-kit/directors/vpc';

export class PrototypeStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
  }

  public async synth(stackName: string, projectName: string) {
    const resourceNamePrefix = [stackName, projectName];
    // Step 1: Get VPC
    const parserVpc = new VpcDirector(PetVpcBuilder).constructVpc(this, 'Vpc');
    // Step 2: Create S3 bucket
    const cdBucket = createTempPrivateBucket([`codedeploy-${stackName}-${projectName}`, this.region, this.account], this);
    // Step 3: Create Roles
    // createGithubCliRole('bpohoriletz', this);
    const cdRole = createCodedeployRole(resourceNamePrefix, cdBucket.bucketArn, this);
    const instanceProfile = createEc2InstanceProfile(resourceNamePrefix, cdBucket.bucketArn, this);
    // Step 4: Create App
    const parserApp = createCdApplication(resourceNamePrefix, this);
    parserApp.applyRemovalPolicy(cdk.RemovalPolicy.DESTROY);
    // Step 5: Create Deployment Group
    const instanceTypes = [new ec2.InstanceType('t2.micro')];
    createPublicDeploymentGroup(resourceNamePrefix, instanceTypes, parserApp, instanceProfile, cdRole, parserVpc, this);
  }
}
