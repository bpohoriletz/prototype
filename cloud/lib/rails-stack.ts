import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { Bucket } from "aws-cdk-lib/aws-s3";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as ssm from "aws-cdk-lib/aws-ssm";

import { createAppRole } from "./aws-cdk-kit/roles/app";
import { createApplication } from "./aws-cdk-kit/eb/app";
import { createEc2InstanceProfile } from "./aws-cdk-kit/roles/ec2-profile";
import { createEnvironment } from "./aws-cdk-kit/eb/env";
import { createInitAppVersions } from "./aws-cdk-kit/eb/app-version";
import { createPrivateBucket } from "./aws-cdk-kit/s3/bucket";
import * as con from "./aws-cdk-kit/naming/resources";

export class RailsStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
  }

  public async synth(stackName: string, projectName: string) {
    const environmentName = "demo";
    const resourceNamePrefix = [stackName, projectName, environmentName];
    // const resourceTags: string[] = [this.stackId, this.region, this.account];
    // Step 0 (tricky): Create regional bucket for ElasticBeanstalk
    const regionalEbBucket = Bucket.fromBucketName(this, "RegionalEbBucket", `elasticbeanstalk-${this.region}-${this.account}`);
    // Step 1: Fetch VPC
    const vpcID = ssm.StringParameter.fromStringParameterName(this, "vpcID", `/${stackName}/VpcID`).stringValue;
    const preProductionVpc = ec2.Vpc.fromLookup(this, "RailsVpc", {vpcName: con.ec2VpcName([stackName])})
    // Step H: Fetch RDS data
    // Step 2: Create ElasticBeanstalk application
    const appRole = createAppRole(resourceNamePrefix, this);
    const preProductionApp = createApplication(resourceNamePrefix, appRole, this);
    // Step 3: Create ElasticBeanstalk environment
    const instanceProfile = createEc2InstanceProfile(resourceNamePrefix, [regionalEbBucket.bucketArn], this);
    const [demoEnv, _demoSg] = await createEnvironment(preProductionApp, resourceNamePrefix, instanceProfile, preProductionVpc, this, "64bit Amazon Linux 2023 v4.4.0 running Ruby 3.4");
    // Step 4 (optional): Create S3 bucket for ElasticBeanstalk environment
    const _ebBucket = createPrivateBucket(resourceNamePrefix, this);
    // Step 5: Create application version
    const appVersion = createInitAppVersions(resourceNamePrefix, preProductionApp, regionalEbBucket.bucketArn, this);
    // Step 6(optional): Deploy application version to ElasticBeanstalk environment
    if (this.node.tryGetContext("deployInitialVersion") == "yes") {
      demoEnv.versionLabel = appVersion.ref;
    }
  }
}
