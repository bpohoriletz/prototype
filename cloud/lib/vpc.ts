import { Stack } from "aws-cdk-lib";
import { Vpc, IpAddresses, SubnetConfiguration, SubnetType } from "aws-cdk-lib/aws-ec2";
import * as con from "./naming/resources";

export function createMinimalVpc(resourceNamePrefix: string[], stack: Stack) : Vpc {
    // configure subnets
    const subnetConfigurations: SubnetConfiguration[] = [
      {
        subnetType: SubnetType.PUBLIC,
        name: con.ec2PublicSubnetName(resourceNamePrefix),
      },
      {
        subnetType: SubnetType.PRIVATE_ISOLATED,
        name: con.ec2IsolatedSubnetName(resourceNamePrefix),
      },
      {
        subnetType: SubnetType.PRIVATE_WITH_EGRESS,
        name: con.ec2PrivateSubnetName(resourceNamePrefix),
      }
    ];
    // create vpc
    const vpc = new Vpc(stack, con.ec2VpcName(resourceNamePrefix), {
      ipAddresses: IpAddresses.cidr("10.0.0.0/16"),
      natGateways: 1,
      subnetConfiguration: subnetConfigurations,
      maxAzs: 2,
    });

    return vpc;
}
