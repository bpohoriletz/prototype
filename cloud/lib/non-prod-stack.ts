import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
// import * as ssm from "aws-cdk-lib/aws-ssm";
// import * as s3 from "aws-cdk-lib/aws-s3";

// import { createNonprodDatabase } from "./aws-cdk-kit/rds/postgres"
import { createVpc } from "./aws-cdk-kit/ec2/vpc";

export class NonProdStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
  }

  public async synth(stackName: string) {
    const resourceNamePrefix = [stackName];
    // Step 1: Create VPC
    const _nonProdVpc = createVpc(resourceNamePrefix, this);
    // Step H: Import ElasticBeanstalk bucket
    // const _sharedVpcId = new ssm.StringParameter(this, `${stackName}VpcID`, {
    //   parameterName: `/${stackName}/VpcID`,
    //   stringValue: nonProdVpc.vpcId,
    //   description: "This parameter stores a configuration value for my application",
    // });
    // Step H: Import ElasticBeanstalk bucket
    // const regionalEbBucket = s3.Bucket.fromBucketName(this, "RegionalEbBucket", `elasticbeanstalk-${this.region}-${this.account}`);
    // Step H: Create RDS server
    // const [_preProductionDb, _rdsSg ] = createNonprodDatabase(resourceNamePrefix, this, preProductionVpc);
  }
}
