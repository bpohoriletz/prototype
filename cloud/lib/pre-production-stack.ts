import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { Vpc } from "aws-cdk-lib/aws-ec2";
import {
  CfnApplication,
  CfnApplicationVersion,
  CfnEnvironment,
} from "aws-cdk-lib/aws-elasticbeanstalk";
import { CfnInstanceProfile, Role } from "aws-cdk-lib/aws-iam";
import { CfnBucket } from "aws-cdk-lib/aws-s3";
import { DatabaseInstance } from "aws-cdk-lib/aws-rds";

import { createMinimalVpc } from "./vpc"
import { createRegionalEbBucket } from "./s3/bucket"
import { createAppRole } from "./roles/app"
import { createApplication } from "./eb/app"
import { createInitAppVersions } from "./eb/app-version"
import { createEc2InstanceProfile } from "./roles/ec2-profile"
import { createEnvironment } from "./eb/env"
import { createDatabase } from "./rds/postgres"

export class PreProductionStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
    const projectName = "prototype";
    const stackName: string = id.replace("Stack", "");
    const environmentName = "demo";
    // dirrerent environments can't be specified with parameters
    // const environmentName: string = new CfnParameter(this, "environmentName", { type: "String", default: "demo" }).valueAsString.toLowerCase();
    const resourceNamePrefix: string[] = [projectName, stackName, environmentName];
    // const resourceTags: string[] = [this.stackId, this.region, this.account];
    // Step 0 (tricky): Create regional bucket for ElasticBeanstalk
    const regionalEbBucket: CfnBucket = createRegionalEbBucket(this);
    // Step 1: Create VPC
    const preProductionVpc: Vpc = createMinimalVpc(resourceNamePrefix, this);
    // Step 2: Create RDS server
    const preProductionDb: DatabaseInstance = createDatabase(resourceNamePrefix, this, preProductionVpc);
    // Step 3: Create ElasticBeanstalk application
    const appRole: Role = createAppRole(resourceNamePrefix, this);
    const preProductionApp: CfnApplication = createApplication(resourceNamePrefix, appRole, this);
    // Step 3: Create ElasticBeanstalk environment
    const instanceProfile: CfnInstanceProfile = createEc2InstanceProfile(resourceNamePrefix, regionalEbBucket.attrArn, this);
    const demoEnv: CfnEnvironment = createEnvironment(preProductionApp, resourceNamePrefix, preProductionVpc, instanceProfile, preProductionDb, this);
    // Step 4 (optional): Create S3 bucket for ElasticBeanstalk environment
    // const demoAppBucket = createPrivateBucket(resourceNamePrefix, this);
    // Step 5: Create application version
    const appVersion: CfnApplicationVersion = createInitAppVersions(resourceNamePrefix, preProductionApp, regionalEbBucket, this)
    // Step 6(optional): Deploy application version to ElasticBeanstalk environment
    if (this.node.tryGetContext("deployInitialVersion") == "yes") {
      demoEnv.versionLabel = appVersion.ref
    }
  }
}
