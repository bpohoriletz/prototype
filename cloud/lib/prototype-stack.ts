import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { Bucket } from "aws-cdk-lib/aws-s3";

import { createAppRole } from "./aws-cdk-kit/roles/app"
import { createApplication } from "./aws-cdk-kit/eb/app"
import { createNonprodDatabase } from "./aws-cdk-kit/rds/postgres"
import { createEc2InstanceProfile } from "./aws-cdk-kit/roles/ec2-profile"
import { createEnvironment } from "./aws-cdk-kit/eb/env"
import { createInitAppVersions } from "./aws-cdk-kit/eb/app-version"
import { createMinimalVpc } from "./aws-cdk-kit/ec2/vpc"
import { createPrivateBucket } from "./aws-cdk-kit/s3/bucket";

export class PrototypeStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
  }

  public async synth(stackName: string, projectName: string) {
    const environmentName = "demo";
    const resourceNamePrefix = [projectName, stackName, environmentName];
    // const resourceTags: string[] = [this.stackId, this.region, this.account];
    // Step 0 (tricky): Create regional bucket for ElasticBeanstalk
    const regionalEbBucket = Bucket.fromBucketName(this, "RegionalEbBucket", `elasticbeanstalk-${this.region}-${this.account}`);
    // Step 1: Create VPC
    const preProductionVpc = createMinimalVpc(resourceNamePrefix, this);
    // Step 2: Create RDS server
    const [_preProductionDb, _rdsSg ] = createNonprodDatabase(resourceNamePrefix, this, preProductionVpc);
    // Step 3: Create ElasticBeanstalk application
    const appRole = createAppRole(resourceNamePrefix, this);
    const preProductionApp = createApplication(resourceNamePrefix, appRole, this);
    // Step 3: Create ElasticBeanstalk environment
    const instanceProfile = createEc2InstanceProfile(resourceNamePrefix, [regionalEbBucket.bucketArn], this);
    const [demoEnv, _demoSg] = await createEnvironment(preProductionApp, resourceNamePrefix, instanceProfile, preProductionVpc, this, "64bit Amazon Linux 2023 v4.4.0 running Ruby 3.4");
    // Step 4 (optional): Create S3 bucket for ElasticBeanstalk environment
    const _ebBucket = createPrivateBucket(resourceNamePrefix, this);
    // Step 5: Create application version
    const appVersion = createInitAppVersions(resourceNamePrefix, preProductionApp, regionalEbBucket.bucketArn, this)
    // Step 6(optional): Deploy application version to ElasticBeanstalk environment
    if (this.node.tryGetContext("deployInitialVersion") == "yes") {
      demoEnv.versionLabel = appVersion.ref
    }
  }
}
