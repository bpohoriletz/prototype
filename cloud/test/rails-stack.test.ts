import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";

import { RailsStack } from "../lib/rails-stack";

describe("new RailsStack()", () => {
  let stack: RailsStack;

  beforeEach(() => {
    const app = new cdk.App();
    const parentStack = new cdk.Stack(app, "StackID");
    stack = new RailsStack(parentStack, "RailsStackID", { env: { account: "00000000", region: "us-east-1"} })
  });

  test("matches snapshot", async () => {
    await stack.synth("stackName", "projectName");

    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});
