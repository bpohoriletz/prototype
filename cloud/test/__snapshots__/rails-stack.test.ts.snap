// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new RailsStack() matches snapshot 1`] = `
{
  "Outputs": {
    "stacknameprojectnamedemoenvApplicationUrl": {
      "Export": {
        "Name": "stackname-projectname-demo-env-application-url",
      },
      "Value": {
        "Fn::GetAtt": [
          "stacknameprojectnamedemoenv",
          "EndpointURL",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-00000000-us-east-1",
          "S3Key": "c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "StackNameProjectNameBctDeploymentAwsCliLayer6B319219",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.11",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-00000000-us-east-1",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-00000000-us-east-1/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::elasticbeanstalk-us-east-1-00000000",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::elasticbeanstalk-us-east-1-00000000/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InitialVersion": {
      "DependsOn": [
        "StackNameProjectNameBctDeploymentAwsCliLayer6B319219",
        "StackNameProjectNameBctDeploymentCustomResourceA1CF8570",
      ],
      "Properties": {
        "ApplicationName": "stackname-projectname-app",
        "Description": "Initial application from AWS",
        "SourceBundle": {
          "S3Bucket": "elasticbeanstalk-us-east-1-00000000",
          "S3Key": "stackname-projectname-app/versions/initial.zip",
        },
      },
      "Type": "AWS::ElasticBeanstalk::ApplicationVersion",
    },
    "OpeningVersion": {
      "DependsOn": [
        "StackNameProjectNameBctDeploymentAwsCliLayer6B319219",
        "StackNameProjectNameBctDeploymentCustomResourceA1CF8570",
      ],
      "Properties": {
        "ApplicationName": "stackname-projectname-app",
        "Description": "First version of the application",
        "SourceBundle": {
          "S3Bucket": "elasticbeanstalk-us-east-1-00000000",
          "S3Key": "stackname-projectname-app/versions/opening.zip",
        },
      },
      "Type": "AWS::ElasticBeanstalk::ApplicationVersion",
    },
    "StackNameProjectNameAppRole6A4FF6FB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudformation.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StackNameProjectNameBctDeploymentAwsCliLayer6B319219": {
      "Properties": {
        "Content": {
          "S3Bucket": "cdk-hnb659fds-assets-00000000-us-east-1",
          "S3Key": "1c5a2ab1d1e53b0672a644454aab3dbb258ccd0079c92ad0e23b95b2c2079f70.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "StackNameProjectNameBctDeploymentCustomResourceA1CF8570": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketKeyPrefix": "stackname-projectname-app/versions",
        "DestinationBucketName": "elasticbeanstalk-us-east-1-00000000",
        "OutputObjectKeys": true,
        "Prune": false,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          "cdk-hnb659fds-assets-00000000-us-east-1",
        ],
        "SourceObjectKeys": [
          "ee3ab666818d5c719930c4d4963a9f4880f341dd6fce9800a1fd5aeda393d727.zip",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "StackNameProjectNameDemoEc2InstProf586C9F3A": {
      "Properties": {
        "InstanceProfileName": "StackNameProjectNameDemoEc2InstProf",
        "Roles": [
          {
            "Ref": "StackNameProjectNameDemoEc2InstRole7FC8C7CB",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "StackNameProjectNameDemoEc2InstRole7FC8C7CB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "StackNameProjectNameDemoLbSgF79314DB": {
      "Properties": {
        "GroupDescription": "Security Group for the Load Balancer",
        "GroupName": "StackNameProjectNameDemoLbSg",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow incoming traffic over port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow incoming traffic over port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "StackNameProjectNameDemoWebSgB07339CD": {
      "Properties": {
        "GroupDescription": "Security Group for the Web instances",
        "GroupName": "StackNameProjectNameDemoWebSg",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "StackNameProjectNameDemoWebSgfromStackIDRailsStackIDStackNameProjectNameDemoLbSg419AAF9C808843543E": {
      "Properties": {
        "Description": "from StackIDRailsStackIDStackNameProjectNameDemoLbSg419AAF9C:80",
        "FromPort": 80,
        "GroupId": {
          "Fn::GetAtt": [
            "StackNameProjectNameDemoWebSgB07339CD",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "StackNameProjectNameDemoLbSgF79314DB",
            "GroupId",
          ],
        },
        "ToPort": 80,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "allowEc2PutStatisticsD9C1A435": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "elasticbeanstalk:PutInstanceStatistics",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "allowEc2PutStatisticsD9C1A435",
        "Roles": [
          {
            "Ref": "StackNameProjectNameDemoEc2InstRole7FC8C7CB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "allowEc2UploadLogsA49BD994": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "allowEc2UploadLogsA49BD994",
        "Roles": [
          {
            "Ref": "StackNameProjectNameDemoEc2InstRole7FC8C7CB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "allowEc2UploadLogsToEbRegionalBucket40230F17": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutObject",
                "s3:ListBucket",
                "s3:ListBucketVersions",
                "s3:GetObject",
                "s3:GetObjectVersion",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::elasticbeanstalk-us-east-1-00000000",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::elasticbeanstalk-us-east-1-00000000/resources/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "allowEc2UploadLogsToEbRegionalBucket40230F17",
        "Roles": [
          {
            "Ref": "StackNameProjectNameDemoEc2InstRole7FC8C7CB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "stacknameprojectnameapp": {
      "Properties": {
        "ApplicationName": "stackname-projectname-app",
        "ResourceLifecycleConfig": {
          "ServiceRole": {
            "Fn::GetAtt": [
              "StackNameProjectNameAppRole6A4FF6FB",
              "Arn",
            ],
          },
          "VersionLifecycleConfig": {
            "MaxCountRule": {
              "DeleteSourceFromS3": true,
              "Enabled": true,
              "MaxCount": 30,
            },
          },
        },
      },
      "Type": "AWS::ElasticBeanstalk::Application",
    },
    "stacknameprojectnamedemoenv": {
      "DependsOn": [
        "stacknameprojectnameapp",
        "StackNameProjectNameDemoEc2InstProf586C9F3A",
      ],
      "Properties": {
        "ApplicationName": "stackname-projectname-app",
        "EnvironmentName": "stackname-projectname-demo-env",
        "OptionSettings": [
          {
            "Namespace": "aws:elasticbeanstalk:application:environment",
            "OptionName": "BUNDLE_PATH",
            "Value": "/var/app/gems",
          },
          {
            "Namespace": "aws:elasticbeanstalk:command",
            "OptionName": "Timeout",
            "Value": "3600",
          },
          {
            "Namespace": "aws:elasticbeanstalk:environment",
            "OptionName": "LoadBalancerType",
            "Value": "application",
          },
          {
            "Namespace": "aws:elasticbeanstalk:healthreporting:system",
            "OptionName": "SystemType",
            "Value": "basic",
          },
          {
            "Namespace": "aws:ec2:instances",
            "OptionName": "InstanceTypes",
            "Value": "t2.micro,t2.small",
          },
          {
            "Namespace": "aws:autoscaling:asg",
            "OptionName": "MaxSize",
            "Value": "4",
          },
          {
            "Namespace": "aws:autoscaling:trigger",
            "OptionName": "BreachDuration",
            "Value": "1",
          },
          {
            "Namespace": "aws:autoscaling:trigger",
            "OptionName": "MeasureName",
            "Value": "TargetResponseTime",
          },
          {
            "Namespace": "aws:autoscaling:trigger",
            "OptionName": "Unit",
            "Value": "Seconds",
          },
          {
            "Namespace": "aws:autoscaling:trigger",
            "OptionName": "UpperThreshold",
            "Value": "9",
          },
          {
            "Namespace": "aws:autoscaling:trigger",
            "OptionName": "LowerThreshold",
            "Value": "0.9",
          },
          {
            "Namespace": "aws:ec2:instances",
            "OptionName": "EnableSpot",
            "Value": "true",
          },
          {
            "Namespace": "aws:ec2:instances",
            "OptionName": "SpotFleetOnDemandBase",
            "Value": "1",
          },
          {
            "Namespace": "aws:ec2:instances",
            "OptionName": "SpotFleetOnDemandAboveBasePercentage",
            "Value": "0",
          },
          {
            "Namespace": "aws:elasticbeanstalk:environment:process:default",
            "OptionName": "HealthCheckPath",
            "Value": "/accounts/sign_in",
          },
          {
            "Namespace": "aws:elasticbeanstalk:environment:process:default",
            "OptionName": "MatcherHTTPCode",
            "Value": "200",
          },
          {
            "Namespace": "aws:autoscaling:launchconfiguration",
            "OptionName": "SecurityGroups",
            "Value": {
              "Fn::GetAtt": [
                "StackNameProjectNameDemoWebSgB07339CD",
                "GroupId",
              ],
            },
          },
          {
            "Namespace": "aws:elbv2:loadbalancer",
            "OptionName": "SecurityGroups",
            "Value": {
              "Fn::GetAtt": [
                "StackNameProjectNameDemoLbSgF79314DB",
                "GroupId",
              ],
            },
          },
          {
            "Namespace": "aws:ec2:vpc",
            "OptionName": "ELBSubnets",
            "Value": "s-12345,s-67890",
          },
          {
            "Namespace": "aws:ec2:vpc",
            "OptionName": "Subnets",
            "Value": "p-12345,p-67890",
          },
          {
            "Namespace": "aws:autoscaling:launchconfiguration",
            "OptionName": "IamInstanceProfile",
            "Value": {
              "Ref": "StackNameProjectNameDemoEc2InstProf586C9F3A",
            },
          },
        ],
        "SolutionStackName": "64bit Amazon Linux 2023 v4.4.0 running Ruby 3.4",
        "Tags": [
          {
            "Key": "Environment",
            "Value": "demo",
          },
          {
            "Key": "Project",
            "Value": "projectName",
          },
        ],
      },
      "Type": "AWS::ElasticBeanstalk::Environment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
